SYNTHESIZER: Add enumerative loop invariant synthesizer

qinheping opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

Dependent on #7429.
This PR is the same as in #7393. But add the enumerative synthesizer into `goto-synthesizer` instead of `goto-instrument`.


- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.


Implement the functionality described below.

# Motivation

This loop invariant synthesizer use the idea counter-example-guided synthesis (CEGIS) to synthesize loop invariants for programs with only checks instrumented by `goto-instrument` with flag `--pointer-check`.

This PR contain the driver of the synthesizer and the verifier we use to check invariant candidates.

# Verifier
The verifier take as input a goto program with pointer checks and a map from loop id to loop invariant candidates. It first annotate and apply the loop invariant candidates into the goto model; and then simulate the CBMC api to verify the instrumented goto program. If there are some violations---loop invariants are not inductive, or some pointer checks fail---, it record valuation from trace generated by the back end to construct a formatted counterexample `cext`.

## Counterexample
A counterexample `cext` record valuations of variables in the trace, and other information about the violation. The valuation we record including
1. set of live variables upon the entry of the loop. 
2. the havoced value of all primitive-typed variables;
3. the havoced offset and the object size of all pointer-typed variables;
4. history values of 2 and 3.

The valuations will be used as true positive (history values) and true negative (havcoed valuation) to filter out bad invariant clause with the idea of the Daikon invariant detector in a following PR. However, in this PR we only construct the valuation but not actually use them.

# Synthesizer
Loop invariants we synthesize are of the form
`` (in_clause || !guard) && (!guard -> pos_clause)``
where `in_clause` and `out_clause` are predicates we store in two different map, and `guard` is the loop guard. The idea is that we separately synthesize the condition `in_clause` that should hold before the execution of the loop body, and condition `pos_clause` that should hold as post-condition of the loop.

When the violation happen in the loop, we update `in_clause`. When the violation happen after the loop, we update `pos_clause`. When the invariant candidate it not inductive, we enumerate strengthening clause to make it inductive.

To be more efficient, we choose different synthesis strategy for different type of violation
* For out-of-boundary violation, we choose to use the violated predicate as the new clause, which is the WLP of the violation if the violation is only dependent on the havocing instruction. **TODO**: to make it more complete, we need to implement a WLP algorithm
* For null-pointer violation, we choose `__CPROVER_same_object(ptr, __CPROVER_loop_entry(ptr))` as the new clause. That is, the havoced pointer should points to the same object at the start of every iteration of the loop. It is a heuristic choice. This can be extended with the idea of alias analysis if needed.
* For invariant-not-preserved violation, we enumerate strengthening clauses and check that if the invariant will be inductive after strengthening (disjunction with the new clause).

The synthesizer works as follow
1. initialize the two invariant maps,
2. verify the current candidates built from the two maps,
_a. return the candidate if there is no violation
_b. synthesize a new clause to resolve the **first** violation and add it to the correct map,
3. repeat 2.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

jimgrundy commented:
Does this depend on PR7127 getting merged?

qinheping commented:
> Does this depend on PR7127 getting merged?

No. It only depend on #7429 which has been merged.

tautschnig commented on regression/goto-synthesizer/chain.sh:
```diff
@@ -41,4 +43,6 @@ elif echo $args_inst | grep -q -- "--dump-c" ; then
 fi
-$cbmc "${name}-mod.gb" ${args_cbmc}
+$goto_synthesizer "${name}-mod.gb" "${name}-mod-2.gb"
+echo "2:"
```
Perhaps this can be turned into some more verbose debugging status note (like "Running CBMC")?

tautschnig commented on regression/goto-synthesizer/loop_contracts_synthesis_02/test.desc:
```diff
@@ -0,0 +1,11 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test shows that loop invariant with form of range predicates can be correctly
```
Should be "a loop invariant" or "loop invariants", and perhaps replace "with form of" by "using"?

tautschnig commented on regression/goto-synthesizer/loop_contracts_synthesis_03/test.desc:
```diff
@@ -0,0 +1,12 @@
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test shows that loop invariant with form of range predicates and same
```
As above: Should be "a loop invariant" or "loop invariants", and perhaps replace "with form of" by "using"?

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+  const goto_programt::const_targett &from,
+  const goto_programt::const_targett &to)
+{
+  std::set<node_indext> visited = std::set<node_indext>();
```
No need for this assignment, this is the default constructor.

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+  std::set<node_indext> visited = std::set<node_indext>();
+  const dep_graph_domaint from_domain = static_cast<const dep_graph_domaint &>(
+    *storage->abstract_state_before(from, *domain_factory));
+  const dep_graph_domaint to_domain = static_cast<const dep_graph_domaint &>(
```
Use references on the left-hand side

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+  const dep_graph_domaint &to,
+  std::set<node_indext> &visited)
+{
+  // `to` is control dependent on `from`?
```
Nit pick: "Is `to` control ..."

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+      return true;
+  }
+
+  // `to` is data dependent on `from`?
```
Nit pick: "Is `to` data dependent .."

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+    if(visited.count((*this)[node].get_node_id()))
+      continue;
+
+    visited.insert((*this)[node].get_node_id());
```
Do this in one operation:
```
if(!visited.insert((*this)[node].get_node_id()).second)
  continue;
```

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+    if(
+      from.get_node_id() == (*this)[node].get_node_id() ||
+      is_flow_dependent(from, (*this)[node], visited))
+      return true;
```
Nit pick: braces around body when the `if` statement has a condition spanning multiple lines.

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+    visited.insert((*this)[node].get_node_id());
+
+    if(
+      from.get_node_id() == (*this)[node].get_node_id() ||
```
There are a lot of uses of `(*this)[node]` here, you might want to introduce a temporary.

tautschnig commented on src/analyses/dependence_graph.cpp:
```diff
@@ -383,6 +383,53 @@ void dependence_grapht::add_dep(
+    if(
+      from.get_node_id() == (*this)[node].get_node_id() ||
+      is_flow_dependent(from, (*this)[node], visited))
+      return true;
```
Same suggestions as for the control-flow case above.

tautschnig commented on src/analyses/dependence_graph.h:
```diff
@@ -167,13 +168,20 @@ class dep_graph_domaint:public ai_domain_baset
+  depst get_data_deps() const
+  {
+    return data_deps;
+  }
```
I suggest that these return const references instead of by value.

tautschnig commented on src/goto-instrument/contracts/contracts.cpp:
```diff
@@ -294,6 +294,12 @@ void code_contractst::check_apply_loop_contracts(
   // Generate havocing code for assignment targets.
   havoc_assigns_targetst havoc_gen(to_havoc, ns);
   havoc_gen.append_full_havoc_code(loop_head_location, pre_loop_head_instrs);
+  // Add loop number to the havoc_code
+  for(auto &havoc_instr : pre_loop_head_instrs.instructions)
+  {
+    if(is_loop_havoc(havoc_instr))
+      havoc_instr.loop_number = loop_end->loop_number;
+  }
```
Can you elaborate as to why this is done? We otherwise only attach loop numbers to (backwards) goto statements. I don't think it would cause a lot of problems, but this seems surprising nonetheless.

qinheping replied:
After loop-contracts transformation, there is no loop and hence no loop number. But in the counter-example-guided synthesis we need locate which loop cause the violation so that we synthesize loop invariant for that loop. My idea is that record original loop number in some instrumented instructions to allow us know which loop the instruction transformed from.
But you are right that using `loop_number` here is ambiguous. I will use a new variable `original_loop_number` with `optionalt<unsigned>` type.

qinheping replied:
As my reply to Daniel said, I will use a map to store the information about original loop numbers in `code_contractst` instead of using `original_loop_number`.

tautschnig commented on src/goto-instrument/contracts/utils.cpp:
```diff
@@ -447,3 +447,36 @@ void generate_history_variables_initialization(
+  if(!target->is_assign())
+    return false;
+
+  return from_expr(target->assign_lhs()).find("__entered_loop") !=
```
Why are you not using the `ENTERED_LOOP` macro here? Also, I don't think you should take a detour via `from_expr` here. Check that the LHS is a symbol_exprt, and do the string containment check on that symbol expression's identifier.

tautschnig commented on src/goto-instrument/contracts/utils.cpp:
```diff
@@ -447,3 +447,36 @@ void generate_history_variables_initialization(
+  if(!target->is_assign())
+    return false;
+
+  if(can_cast_expr<symbol_exprt>(target->assign_lhs()))
+  {
+    const auto &lhs = to_symbol_expr(target->assign_lhs());
+    return id2string(lhs.get_identifier()).find("::" + var_name) !=
+           std::string::npos;
+  }
```
That's pretty much the code I'd expect in `is_transformed_loop_end` instead of `from_expr`

tautschnig commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -16,7 +16,14 @@ Date: July 2021
+{
+  if(id2string(instruction.source_location().get_comment()) == IS_LOOP_HAVOC)
+    return true;
+  return false;
```
How about `return id2string(....`?

tautschnig commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -52,6 +59,8 @@ void havoc_utilst::append_object_havoc_code_for_expr(
   havoc.add_source_location() = location;
   havoc.add_to_operands(expr);
   dest.add(goto_programt::make_other(havoc, location));
+  dest.instructions.back().source_location_nonconst().set_comment(
+    IS_LOOP_HAVOC);
```
How hard would it be to reverse-engineer that a particular instruction is a loop-havoc from the instruction structure instead of using this comment-based marking?

qinheping replied:
A loop-havoc is an assignment with arbitrary lhs and `side_effect_expr_nondett` rhs. Such assignments are also used in function contracts: https://github.com/diffblue/cbmc/blob/develop/src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp#L171.
I am afraid we cannot decide whether an assignment is loop-havoc from only its structure.

tautschnig commented on src/goto-programs/loop_ids.h:
```diff
@@ -31,14 +31,28 @@ struct loop_idt
   {
   }
 
+  loop_idt() : function_id(""), loop_number()
+  {
+  }
+
```
Where do you actually use this one (which piece of code would no longer compile if you omitted it)?

qinheping replied:
When we compute the cause loop id, the loop number without value indicates that no cause loop found. This constructor constructs a loop id with `loop_number` having no value.

https://github.com/diffblue/cbmc/blob/f5629d636c7e9b2db4f7e8f8851b74bca12b4fce/src/goto-synthesizer/cegis_verifier.cpp#L480

tautschnig replied:
Hmm, yes, so I think this reinforces [the comment](https://github.com/diffblue/cbmc/pull/7430#discussion_r1048112384) that I provided in that function: I believe that function should return an `optionalt<loop_idt>` rather than (possibly) returning a not-really-initialised `loop_idt`.

tautschnig commented on src/goto-programs/loop_ids.h:
```diff
@@ -31,14 +31,28 @@ struct loop_idt
+  loop_idt(const loop_idt &other)
+    : function_id(other.function_id), loop_number(other.loop_number)
+  {
+  }
```
`loop_idt(const loop_idt &other) = default;` will do

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  {
+    if(contains_symbol_prefix(*it, prefix))
+      return true;
+  }
```
I'd suggest you avoid the recursion by doing:
```
for(auto it = expr.depth_begin(), itend = expr.depth_end();
      it != itend; ++it)
{
  if(
    expr.id() == ID_symbol &&
    has_prefix(id2string(to_symbol_expr(expr).get_identifier()), prefix))
  {
    return true;
  }
}

return false;
```

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  options.set_option("symex-cache-dereferences", false);
+  options.set_option("rewrite-union", true);
+  options.set_option("self-loops-to-assumptions", true);
+  options.set_option("trace", true);
```
I don't think this is the default?

qinheping replied:
The first three are not set in `set_default_options`, but are set in `cbmc_parse_optionst::get_command_line_options`, assuming no command lines are given when running cbmc.
https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_parse_options.cpp#L303
https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_parse_options.cpp#L364
https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_parse_options.cpp#L268

The last one if for generating traces such that we can construct counterexamples from traces.

qinheping replied:
However, all tests pass even the first three are removed.
So I am OK to remove them.

tautschnig replied:
My apologies, I think my comment was confusing: I only meant to refer to the very last one ("trace") as not being a default (despite the comment at the beginning of the block suggesting this. So really, all I meant to ask for is tweak the comment to make clear which ones are the default vs which one you've specifically added.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  options.set_option("self-loops-to-assumptions", true);
+  options.set_option("trace", true);
+
+  // Preporcess `goto_model`. Copied from `cbmc_parse_options.cpp`.
```
s/Preporcess/Preprocess/

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  const goto_tracet &goto_trace,
+  const goto_programt::const_targett violation)
+{
+  loop_idt result;
```
Ah, this is where you currently use the default constructor. How about using `optionalt<loop_idt>` here?

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+      // `from` : loop havocing
+      // `to`   : violation
+      goto_programt::const_targett from;
+      goto_programt::const_targett to;
```
Uninitialised iterators are scary. Use `optionalt` here.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+      // if it is dependent on the loop havoc.
+      if(dependence_graph.is_flow_dependent(from, to))
+      {
+        // TODO: not safe for inlined static functions.
```
Can you elaborate? Perhaps also sketch the plan for how to resolve this?

qinheping replied:
The problem is the function name returned by `get_function()` is not the key we use to store the function in the function map of the function model.
My plan is to use a function to decide whether the function name we get is of an inlined static function. If yes, we rewrite the name to the form of keys of function maps.

tautschnig replied:
Would `step.function_id` (instead of `step.pc->source_location().get_function()`) solve the problem for you?

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+      if(dependence_graph.is_flow_dependent(from, to))
+      {
+        // TODO: not safe for inlined static functions.
+        result.function_id = step.pc->source_location().get_function();
+        result.loop_number = step.pc->loop_number;
```
What if you ended in here multiple times? You are returning one result only, so should you contemplate returning early?

qinheping replied:
This is one limitation of the current implementation---we only synthesize loop contracts for one loop at a time. So when there are multiple cause loops, we choose to synthesize loop contracts for the closest one---the last we we entered in the trace.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+
+  // Loop-entry valuations of havoced variables.
+  std::map<exprt, std::size_t> loop_entry_values;
+  std::map<exprt, std::size_t> loop_entry_offsets;
```
I'm pretty sure you want `std::unordered_map` here. We even cache hash values to speed this up.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  std::map<exprt, std::size_t> loop_entry_offsets;
+
+  // Live variables upon the loop head.
+  std::set<exprt> live_variables;
```
std::unordered_set

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+        // Live variables
+        // 1. must be in the same function as the target loop;
+        // 2. alive before entering the target loop;
+        // 3. a pointer or a primitive-typed variable;
+        // TODO: add support for union pointer
+        if(
+          step.pc->source_location().get_function() ==
+            loop_entry_loc.get_function() &&
+          !entered_loop &&
+          (step.full_lhs.type().id() == ID_unsignedbv ||
+           step.full_lhs.type().id() == ID_pointer) &&
```
Can you elaborate why exactly this subset of types is the right one?

qinheping replied:
They are the type of variables we enumerate in the synthesized invariants. That is, we only enumerate expressions containing pointer types and unsigned primitive types in the current implementation.
I wouldn't say this subset is the right one. This subset is the least one we need to synthesize loop invariants for aws-c-common and coreJSON benchmarks. Because memory safety is about pointer, size, and offset, I would say this set is a reasonable set of types we consider when we are synthesizing loop invariants for memory safety. 
Maybe `signedbv` should also be included.

tautschnig replied:
+1 to adding `signedbv` in here.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+          can_cast_expr<annotated_pointer_constant_exprt>(
+            step.full_lhs_value) &&
+          contains_symbol_prefix(
+            step.full_lhs_value, SYMEX_DYNAMIC_PREFIX "dynamic_object"))
```
As of d2b56c21f6b30 you'll have to use `"::dynamic_object"` (i.e., insert a double colon).

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  const namespacet ns(goto_model.symbol_table);
+
+  // Valuations of havoced variables right after havoc instructions.
+  std::map<exprt, std::size_t> object_sizes;
+  std::map<exprt, std::size_t> havoced_values;
+  std::map<exprt, std::size_t> havoced_pointer_offsets;
```
Here and below: is `std::size_t` the right type here? I'd suspect that these should be `mp_integer`.

qinheping replied:
They should cover the range of pointer offsets, or user-defined primitive-type integers. `std::size_t` could be not enough.
I will use `mp_integer` instead.

tautschnig replied:
As the guideline on this: use `std::size_t` when you are limited to the host platform running CBMC. Use `mp_integer` whenever talking about the data domain of the verification target (which might be a platform entirely different from the host).

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+          if(!entered_loop)
+          {
+            loop_entry_values[step.full_lhs] =
+              hex_to_size_t(step.full_lhs_value.get_string(ID_value));
```
The fact that we use a hexadecimal encoding at present is an implementation detail. Please use `bvrep2integer` instead, or simply `numeric_cast_v<std::size_t>(to_constant_expr(step.full_lhs))` (though it'll be `<mp_integer>` instead of `<std::size_t>` when taking into account my comment above).

qinheping replied:
I use `bvrep2integer` now.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+          if(is_loop_havoc(*step.pc))
+          {
+            havoced_values[step.full_lhs] =
+              hex_to_size_t(step.full_lhs_value.get_string(ID_value));
```
As above. You are really don't want `hex_to_size_t`.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  // 2. run the CBMC API to verify the intrumented goto model. As the API is not
+  //    merged yet, we preprocess the goto model and run the symex checker on it
+  //    to simulate CBMC API.
+  //    FIXEME: ^^^ replace the symex checker once the real API is merged.
```
s/FIXEME/TODO/ to have something more consistent to grep for.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,524 @@
+  log.get_message_handler().set_verbosity(original_verbosity);
+
+  //
+  // Strat to counstruct the counterexample.
```
s/Strat/Start/, s/counstruct/construct/

tautschnig commented on src/goto-synthesizer/cegis_verifier.h:
```diff
@@ -0,0 +1,142 @@
+  {
+  }
+
+  cext()
+  {
+  }
```
Safe to omit

tautschnig replied:
Seems this still needs to be removed?

qinheping replied:
I tried to omit it. But failed to compile with error 
```
/Users/qinhh/Repos/CBMC/clone_2/src/goto-synthesizer/cegis_verifier.h:90:3: error: constructor for 'cegis_verifiert' must explicitly initialize the member 'return_cex' which does not have a default constructor
  cegis_verifiert(
  ^
/Users/qinhh/Repos/CBMC/clone_2/src/goto-synthesizer/cegis_verifier.h:105:8: note: member is declared here
  cext return_cex;
       ^
/Users/qinhh/Repos/CBMC/clone_2/src/goto-synthesizer/cegis_verifier.h:25:7: note: 'cext' declared here
class cext
      ^
1 error generated.
```

tautschnig replied:
Perhaps the use should be fixed? A default-constructed `cext` seems ill-formed. Perhaps that `return_cex` class member shouldn't exist?

qinheping replied:
Agreed, the return value of `verify()` is now `optionalt<cext>`---it return a counterexample if there is some violation.

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.h:
```diff
@@ -44,7 +44,38 @@ class enumerative_loop_contracts_synthesizert
+  invariant_mapt neg_guards;
+
+  /// Map from tmp_post variables to their original variables.
+  std::map<exprt, exprt> tmp_post_map;
```
Likely `std::unordered_map`

tautschnig commented on src/goto-synthesizer/expr_enumerator.cpp:
```diff
@@ -316,7 +316,18 @@ exprt binary_functional_enumeratort::instantiate(const expr_listt &exprs) const
   if(op_id == ID_equal)
+  {
+    if(exprs.front().type() != exprs.back().type())
+      return true_exprt();
```
Why is this the right thing to do?

qinheping replied:
It is safe to do this. But it is not a perfect fix for the problem
The problem is that unsignedbv types with different width are actually different types. When build the grammar for the enumerator, I put all unsignedbv variables as terminals for one non-terminal. Then the grammar could enumerate equal expression between two variables with different types. I let it enumerate true here to avoid this happen.
Example
N -> N == N | N + N | `a` | `b` | `c`
where `a` and `b` are of `unsigned` type and `c` is of `unsigned long`.
I am thinking about building well-typed grammar. But maybe a better fix here is to widen the short unsignedbv to the long one using type cast. What do you think?

tautschnig replied:
If it's strictly about `unsigned(bv)`, then widening via typecast seems safe.

tautschnig commented on src/goto-synthesizer/cegis_verifier.cpp:
```diff
@@ -0,0 +1,529 @@
+        from.has_value() && to.has_value() &&
+        dependence_graph.is_flow_dependent(from.value(), to.value()))
+      {
+        result =
+          loop_idt(step.function_id, step.pc->original_loop_number.value());
```
As a follow-up to the earlier discussion: as you do one loop at a time, would it be equally ok to do the first loop? If so, you could return early and save some compute time.

tautschnig commented on src/goto-programs/loop_ids.h:
```diff
@@ -31,6 +31,10 @@ struct loop_idt
   {
   }
 
+  loop_idt() = delete;
```
I don't think you need this line.

tautschnig commented on src/goto-synthesizer/synthesizer_utils.cpp:
```diff
@@ -112,3 +116,55 @@ void annotate_invariants(
+  for(const auto &in_clause : in_clauses)
+  {
+    const auto &id = in_clause.first;
+    INVARIANT(neg_guards.count(id), "Some loop guard is missing.");
+
+    result[id] = and_exprt(
+      or_exprt(neg_guards.at(id), in_clause.second),
+      implies_exprt(neg_guards.at(id), post_clauses.at(id)));
```
That's three lookups in `neg_guards`. Perhaps you can use a single `find`, and make the `INVARIANT` test that the resulting iterator is not `end()`.

tautschnig commented on src/goto-synthesizer/synthesizer_utils.h:
```diff
@@ -50,7 +53,20 @@ get_loop_head(const unsigned int loop_number, goto_functiont &function);
+  const goto_functiont &function,
+  unsigned location_number_of_target);
+
+///  Combine invariant of form
+///   (in_inv || !guard) && (!guard -> pos_inv)
```
Nit pick: too much whitespace after the `///`

tautschnig commented on src/goto-synthesizer/synthesizer_utils.h:
```diff
@@ -9,6 +9,9 @@ Author: Qinheping Hu
 #ifndef CPROVER_GOTO_SYNTHESIZER_SYNTHESIZER_UTILS_H
 #define CPROVER_GOTO_SYNTHESIZER_SYNTHESIZER_UTILS_H
 
+#include <util/pointer_expr.h>
```
It's not obvious why this include is newly needed?

qinheping replied:
I added it for fixing some CI failures. Let me remove it now and see what CI says.

tautschnig commented on src/goto-synthesizer/synthesizer_utils.cpp:
```diff
@@ -112,3 +116,55 @@ void annotate_invariants(
+  // loop end of transformed code.
+
+  bool result = false;
+  unsigned location_number_of_havocing = UINT_MAX;
```
Use `optionalt` instead of a magic constant.

tautschnig commented on src/goto-synthesizer/synthesizer_utils.cpp:
```diff
@@ -112,3 +116,55 @@ void annotate_invariants(
+
+    if(is_transformed_loop_end(it) && it->loop_number == loop_id.loop_number)
+    {
+      result =
+        result || (location_number_of_havocing < location_number_of_target &&
+                   location_number_of_target < it->location_number);
```
Can't you just do `if(....) return true;`? Once `result` has become `true`, the remainder of the expression will never be evaluated again.

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -32,19 +92,275 @@ void enumerative_loop_contracts_synthesizert::init_candidates()
+    }
+    else
+    {
+      // violation happens pos-loop.
```
"post-loop"?

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -32,19 +92,275 @@ void enumerative_loop_contracts_synthesizert::init_candidates()
+    dependent_symbols = compute_dependent_symbols(
+      cause_loop_id, new_invariant_clause, verifier.return_cex.live_variables);
+
+    // add the new cluase to the candidate invairants
```
s/invairants/invariants/

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -11,8 +11,68 @@ Author: Qinheping Hu
+  }
+}
+
+std::vector<exprt> construct_terminals(std::set<symbol_exprt> symbols)
```
You will want to pass a const reference here.

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -11,8 +11,68 @@ Author: Qinheping Hu
+      // __CPROVER_pointer_offset(v),
+      // __CPROVER_pointer_offset(__CPROVER_loop_entry(v))
+      // into the result.
+      result.push_back(unary_exprt(ID_pointer_offset, e, size_type()));
```
Is there a reason not to use `pointer_offset_exprt` here?

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -11,8 +11,68 @@ Author: Qinheping Hu
+
+std::vector<exprt> construct_terminals(std::set<symbol_exprt> symbols)
+{
+  std::vector<exprt> result = std::vector<exprt>();
```
No need for the assignment.

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -11,8 +11,68 @@ Author: Qinheping Hu
+      substitute_tmp_post_rec(*it, tmp_post_map);
+
+  // possibly substitute?
+  if(dest.id() == ID_symbol && tmp_post_map.count(dest))
+  {
+    dest = tmp_post_map.at(dest);
```
Use `find`to avoid the duplicate lookup in `tmp_post_map`

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -11,8 +11,68 @@ Author: Qinheping Hu
+#include "cegis_verifier.h"
+#include "expr_enumerator.h"
+
+// substitute all tmp_post variables with their origins in `expr`
+void substitute_tmp_post_rec(
+  exprt &dest,
+  const std::unordered_map<exprt, exprt, irep_hash> &tmp_post_map)
+{
+  if(dest.id() != ID_address_of)
+    Forall_operands(it, dest)
+      substitute_tmp_post_rec(*it, tmp_post_map);
+
+  // possibly substitute?
+  if(dest.id() == ID_symbol && tmp_post_map.count(dest))
+  {
+    dest = tmp_post_map.at(dest);
+  }
+}
```
I believe you could get rid of this and use `replace_symbolt` instead.

kroening commented on src/goto-programs/goto_program.h:
```diff
@@ -541,6 +541,10 @@ class goto_programt
+    /// We record the original loop number for some instrumended instructions
+    /// so that we know which loop they are transformed from.
+    optionalt<unsigned> original_loop_number;
+
```
Please drop, this is not needed, and costs extra memory.

qinheping replied:
Sure. I plan to instead define `ID_original_loop_number` and store the original loop number in the field `ID_original_loop_number` of the lhs `exprt` of instrumented assignments such as loop-havoc. Does it sound OK to you?

kroening replied:
No, please use a map.

qinheping replied:
Thank you, I will store the map from instrumented instructions to their original loop number in `code_contractst`. And the functions I used to decided whether an instruction is a loop-havoc or instrumented instruction will be no longer needed with such map.

tautschnig reviewed (changes requested):
Adding a blocker just to make sure we don't end up inadvertently merging this: must not be merged until the change in goto_program.h that was added after my earlier approval is removed (cf. conversation with Daniel).

tautschnig commented on src/ansi-c/compiler_headers/arm_builtin_headers.inc:
```diff
@@ -0,0 +1,45 @@
+"// clang-format off\n"
+"void __breakpoint(int val);\n"
+"void __cdp(unsigned int coproc, unsigned int ops, unsigned int regs);\n"
+"void __clrex(void);\n"
+"unsigned char __clz(unsigned int val);\n"
+"unsigned int __current_pc(void);\n"
```
Something has gone wrong here, this is an auto-generated file. (There are a few others of this kind added in the latest commit.)

qinheping commented:
> Adding a blocker just to make sure we don't end up inadvertently merging this: must not be merged until the change in goto_program.h that was added after my earlier approval is removed (cf. conversation with Daniel).

I added another commit to remove undo the changes in `goto_program.h`. I use a map to store the original loop numbers for some instrumented instructions in `code_contractst`.
@kroening

kroening commented:
> I added another commit to remove undo the changes in `goto_program.h`. I use a map to store the original loop numbers for some instrumented instructions in `code_contractst`.

Please squash the commits!

martin-cs commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -16,7 +16,13 @@ Date: July 2021
 #include <util/std_code.h>
 
-#include <goto-programs/goto_program.h>
+#define IS_LOOP_HAVOC "Loop havocing instrumented by applying loop contracts."
```
This makes me nervous.  At very least could we have it as a string in irep_ids?

martin-cs commented on src/goto-instrument/havoc_utils.cpp:
```diff
@@ -62,4 +70,6 @@ void havoc_utilst::append_scalar_havoc_code_for_expr(
   dest.add(goto_programt::make_assignment(
     code_assignt{expr, std::move(rhs), location}, location));
+  dest.instructions.back().source_location_nonconst().set_comment(
+    IS_LOOP_HAVOC);
```
I appreciate that you want to put some kind of tag in the program but this feels fragile and ad-hoc.  Would it be possible to use a function call to `end_of_loop_havoc_block` or a `LOCATION` or even an `OTHER` instruction?  Without knowing the wider context (hard to get from such a big commit) I can't really say what is best.

qinheping replied:
I removed `is_loop_havoc` function and related codes.
Instead, I changed the structure of transformed loop from
```
havoc(write_set)
```
to
```
DECL __is_in_loop_havoc_block
ASSIGN __is_in_loop_havoc_block = true
havoc(write_set)
ASSIGN __is_in_loop_havoc_block = false
```
Now, after applying a loop contract, `code_cotractst` will scan instrumented instructions to find all assignments in loop-havoc blocks and store them into a set `loop_havoc_set`.

@tautschnig : We cannot decide whether an assignment is loop havoc from its  instruction structure. So I change the structure of transformed loops to better identify loop havoc.

martin-cs reviewed (changes requested):
Hi @qinheping , thanks for this; it is a really great piece of work and it is long-overdue that we have this capability.  Given that, I feel a bit bad critiquing it.  I have only looked at the code in `analyses` and `goto-instrument` (but not contracts).

As I understand it, the functionality you want for dependence-graphs is to be able to do reachability between two program locations.  Is this correct?  If so, `dependence_grapht` inherits from `grapht` so you should be able to use the graph functions directly.  I would far rather we added an `can_reach` method to `grapht` than custom searching here because it decouples the computation of the dependence graph and the use of it.  There is another set of dependency graph code `analyses/variable-sensitivity/variable_sensitivity_dependence_graph.{h,cpp}` which is more configurable and hopefully more precise, it would be nice to be able to switch between them if there are issues.

Perhaps I have misunderstood as your `is_flow_dependent` seems to consider control and then data, but doesn't then check control dependency again?  If you want this behaviour then maybe use the graph visitor rather than recursing?

Also, non-blocking thoughts:
- The commit message is great; can that make it in to the documentation?
- I appreciate that in this case it is difficult but it would help reviewing in the future if things could be broken out a bit more.

qinheping commented:
> Hi @qinheping , thanks for this; it is a really great piece of work and it is long-overdue that we have this capability. Given that, I feel a bit bad critiquing it. I have only looked at the code in `analyses` and `goto-instrument` (but not contracts).
> 
> As I understand it, the functionality you want for dependence-graphs is to be able to do reachability between two program locations. Is this correct? If so, `dependence_grapht` inherits from `grapht` so you should be able to use the graph functions directly. I would far rather we added an `can_reach` method to `grapht` than custom searching here because it decouples the computation of the dependence graph and the use of it. There is another set of dependency graph code `analyses/variable-sensitivity/variable_sensitivity_dependence_graph.{h,cpp}` which is more configurable and hopefully more precise, it would be nice to be able to switch between them if there are issues.
> 
> Perhaps I have misunderstood as your `is_flow_dependent` seems to consider control and then data, but doesn't then check control dependency again? If you want this behaviour then maybe use the graph visitor rather than recursing?
> 
> Also, non-blocking thoughts:
> 
>     * The commit message is great; can that make it in to the documentation?
> 
>     * I appreciate that in this case it is difficult but it would help reviewing in the future if things could be broken out a bit more.

Thank you! I removed all the changes to `dependence_grapht` and now use the `get_reachable` function instead.

qinheping committed:
Add enumerative loop invariant synthesizer

Implement the functionality described below.

Motivation
---
This loop invariant synthesizer use the idea counter-example-guided synthesis (CEGIS) to synthesize loop invariants for programs with only checks instrumented by `goto-instrument` with flag `--pointer-check`.

This PR contain the driver of the synthesizer and the verifier we use to check invariant candidates.

Verifier
---
The verifier take as input a goto program with pointer checks and a map from loop id to loop invariant candidates. It first annotate and apply the loop invariant candidates into the goto model; and then simulate the CBMC api to verify the instrumented goto program. If there are some violations---loop invariants are not inductive, or some pointer checks fail---, it record valuation from trace generated by the back end to construct a formatted counterexample `cext`.

Counterexample
---
A counterexample `cext` record valuations of variables in the trace, and other information about the violation. The valuation we record including
1. set of live variables upon the entry of the loop.
2. the havoced value of all primitive-typed variables;
3. the havoced offset and the object size of all pointer-typed variables;
4. history values of 2 and 3.

The valuations will be used as true positive (history values) and true negative (havcoed valuation) to filter out bad invariant clause with the idea of the Daikon invariant detector in a following PR. However, in this PR we only construct the valuation but not actually use them.

Synthesizer
---
Loop invariants we synthesize are of the form
`` (in_clause || !guard) && (!guard -> pos_clause)``
where `in_clause` and `out_clause` are predicates we store in two different map, and `guard` is the loop guard. The idea is that we separately synthesize the condition `in_clause` that should hold before the execution of the loop body, and condition `pos_clause` that should hold as post-condition of the loop.

When the violation happen in the loop, we update `in_clause`. When the violation happen after the loop, we update `pos_clause`. When the invariant candidate it not inductive, we enumerate strengthening clause to make it inductive.

To be more efficient, we choose different synthesis strategy for different type of violation
* For out-of-boundary violation, we choose to use the violated predicate as the new clause, which is the WLP of the violation if the violation is only dependent on the havocing instruction. **TODO**: to make it more complete, we need to implement a WLP algorithm
* For null-pointer violation, we choose `__CPROVER_same_object(ptr, __CPROVER_loop_entry(ptr))` as the new clause. That is, the havoced pointer should points to the same object at the start of every iteration of the loop. It is a heuristic choice. This can be extended with the idea of alias analysis if needed.
* For invariant-not-preserved violation, we enumerate strengthening clauses and check that if the invariant will be inductive after strengthening (disjunction with the new clause).

The synthesizer works as follow
1. initialize the two invariant maps,
2. verify the current candidates built from the two maps,
_a. return the candidate if there is no violation
_b. synthesize a new clause to resolve the **first** violation and add it to the correct map,
3. repeat 2.

martin-cs reviewed (approved):
@qinheping Thank you for the quick and accommodating changes.  I am approving this although I don't think it touches anywhere I am a maintainer any more.  Thanks for all your work.
